pkg/controllers/management/cis/scheduled_scan.go

```diff
@@ -0,0 +1,272 @@
+package cis
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/rancher/rancher/pkg/controllers/user/cis"
+	managementv3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	DefaultCronSchedule    = "0 0 * * *"
+	DefaultRetention       = 3
+	ScheduledScanKeyPrefix = "scheduledscan/"
+	separator              = ":"
+)
+
+type scheduleInfo struct {
+	cronSchedule       string
+	lastScanLaunchedAt time.Time
+	nextScanAt         time.Time
+}
+
+type scheduledScanHandler struct {
+	sync.Mutex
+	managementClient  managementv3.Interface
+	clusterClient     v3.ClusterInterface
+	clusterLister     v3.ClusterLister
+	clusterScanClient v3.ClusterScanInterface
+	clusterScanLister v3.ClusterScanLister
+	clustersMap       map[string]*scheduleInfo
+}
+
+func newScheduleScanHandler(
+	managementClient managementv3.Interface,
+	clusterClient v3.ClusterInterface,
+	clusterLister v3.ClusterLister,
+	clusterScanClient v3.ClusterScanInterface,
+	clusterScanLister v3.ClusterScanLister,
+) *scheduledScanHandler {
+	return &scheduledScanHandler{
+		managementClient:  managementClient,
+		clusterClient:     clusterClient,
+		clusterLister:     clusterLister,
+		clusterScanClient: clusterScanClient,
+		clusterScanLister: clusterScanLister,
+		clustersMap:       make(map[string]*scheduleInfo),
+	}
+}
+
+func (i *scheduleInfo) String() string {
+	return fmt.Sprintf("cronSchedule: %v lastAt: %v nextAt: %v",
+		i.cronSchedule, i.lastScanLaunchedAt.String(), i.nextScanAt.String())
+}
+
+func (ssh *scheduledScanHandler) clusterSync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if cluster == nil ||
+		cluster.DeletionTimestamp != nil ||
+		cluster.Spec.ScheduledClusterScan == nil {
+		return nil, nil
+	}
+
+	ssh.Lock()
+	defer ssh.Unlock()
+	clusterInfo, ok := ssh.clustersMap[cluster.Name]
+	if cluster.Spec.ScheduledClusterScan.Enabled {
+		if !ok {
+			// Do stuff to enable scheduling
+			cronSchedule := getCronSchedule(cluster)
+			clusterInfo = &scheduleInfo{
+				cronSchedule: cronSchedule,
+			}
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan enabled for cluster %v: %v",
```

Comment:
do we want to log this at debug?

Reply:
This gets printed only when it's enabled, not always, so should be ok.